src/Storage/Database.php

```diff
@@ -107,20 +107,31 @@ public function getExpired($amount)
         return $q->execute()->fetchAll(\PDO::FETCH_COLUMN);
     }
 
-    public function getByTags(array $tags)
+    public function getByTags(array $tags): array
     {
         if (!$tags) {
             return [];
         }
 
         $q = $this->db->select(self::TABLE_ENTRIES, 'c')
-            ->fields('c', ['id']);
+            ->fields('c', []);
         $q->condition('c.expiry', time(), '>=');
         $q->innerJoin(self::TABLE_TAGS, 't', 't.id = c.id');
         $q->condition('t.tag', $tags, 'IN');
 
-        $ids = $q->execute()->fetchAll(\PDO::FETCH_COLUMN);
-        return $ids;
+        return array_map(
+            static function (array $data) {
+                return new Cache(
+                    $data['id'],
+                    $data['uri'],
+                    $data['method'],
+                    $data['content'],
+                    unserialize($data['headers'], ['allowed_classes' => false]),
+                    $data['expiry']
+                );
+            },
+            $q->execute()->fetchAll(\PDO::FETCH_ASSOC)
+        );
```

Comment:
Don't load all content here. When a cache tag that is shown on all pages is invalidated you are now loading all pages in memory